File: packages/python/linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList:
    def __init__(self):
        self.head = None
    
    def insert(self, data):
        if not self.head:
            self.head = Node(data)
        else:
            cur = self.head
            while cur.next:
                cur = cur.next
            cur.next = Node(data)
            
    def display(self):
        cur = self.head
        res = []
        while cur:
            res.append(cur.data)
            cur = cur.next
        return res

    def reverse(self):
        prev = None
        cur = self.head
        while cur:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt
        self.head = prev

    def has_cycle(self):
        slow = self.head
        fast = self.head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False

class DNode:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None
        
    def insert(self, data):
        if not self.head:
            self.head = DNode(data)
        else:
            cur = self.head
            while cur.next:
                cur = cur.next
            new_node = DNode(data)
            cur.next = new_node
            new_node.prev = cur
            
    def display(self):
        cur = self.head
        res = []
        while cur:
            res.append(cur.data)
            cur = cur.next
        return res

File: packages/c/LinkedList.c
#include "LinkedList.h"
#include <stdio.h>
#include <stdlib.h>

SNode* insert_singly(SNode* head, int val) {
    SNode* node = (SNode*)malloc(sizeof(SNode));
    node->data = val;
    node->next = head;
    return node;
}
void print_singly(SNode* head) {
    while (head) {
        printf("%d ", head->data);
        head = head->next;
    }
    printf("\n");
}
void free_singly(SNode* head) {
    while (head) {
        SNode* temp = head;
        head = head->next;
        free(temp);
    }
}

DNode* insert_doubly(DNode* head, int val) {
    DNode* node = (DNode*)malloc(sizeof(DNode));
    node->data = val;
    node->next = head;
    node->prev = NULL;
    if (head) {
        head->prev = node;
    }
    return node;
}
void print_doubly(DNode* head) {
    while (head) {
        printf("%d ", head->data);
        head = head->next;
    }
    printf("\n");
}
void free_doubly(DNode* head) {
    while (head) {
        DNode* temp = head;
        head = head->next;
        free(temp);
    }
}

SNode* reverse(SNode* head) {
    SNode* prev = NULL;
    SNode* current = head;
    while (current != NULL) {
        SNode* next = current->next;
        current->next = prev;
        prev = current;
        current = next;
    }
    return prev;
}

int has_cycle(SNode* head) {
    if (!head || !head->next) return 0;
    SNode* slow = head;
    SNode* fast = head;
    while (fast != NULL && fast->next != NULL) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) return 1;
    }
    return 0;
}